phoenix-core/src/main/java/org/apache/phoenix/schema/types/PJsonDataType.java

```diff
@@ -0,0 +1,258 @@
+ * as the order of keys within JSON objects. Also, if a JSON object within the
+ * value contains the same key more than once, all the key/value pairs are kept.
+ * It stores the data as string in single column of HBase and it has same data
+ * size limit as Phoenix's Varchar.
+ * <p>
+ * JSON data types are for storing JSON (JavaScript Object Notation) data, as
+ * specified in RFC 7159. Such data can also be stored as text, but the JSON
+ * data types have the advantage of enforcing that each stored value is valid
+ * according to the JSON rules.
+ */
+public class PJsonDataType extends PDataType<String> {
+
+	public static final PJsonDataType INSTANCE = new PJsonDataType();
+
+	PJsonDataType() {
+		super("JSON", Types.OTHER, PhoenixJson.class, null, 48);
+	}
+
+	@Override
+	public int toBytes(Object object, byte[] bytes, int offset) {
+
+		if (object == null) {
+			return 0;
+		}
+		byte[] b = toBytes(object);
+		System.arraycopy(b, 0, bytes, offset, b.length);
+		return b.length;
+
+	}
+
+	@Override
+	public byte[] toBytes(Object object) {
+		if (object == null) {
+			return ByteUtil.EMPTY_BYTE_ARRAY;
+		}
+		PhoenixJson phoenixJson = (PhoenixJson) object;
+		return PVarchar.INSTANCE.toBytes(phoenixJson.toString());
+	}
+
+	@Override
+	public Object toObject(byte[] bytes, int offset, int length,
+			@SuppressWarnings("rawtypes") PDataType actualType,
+			SortOrder sortOrder, Integer maxLength, Integer scale) {
+
+		if (!actualType.isCoercibleTo(this)) {
+			throwConstraintViolationException(actualType, this);
+		}
+		if (length == 0) {
+			return null;
+		}
+		return getPhoenixJson(bytes, offset, length);
+
+	}
+
+	@Override
+	public Object toObject(Object object,
+			@SuppressWarnings("rawtypes") PDataType actualType) {
+		if (object == null) {
+			return null;
+		}
+		if (equalsAny(actualType, PJsonDataType.INSTANCE)) {
+			return object;
+		}
+		if (equalsAny(actualType, PVarchar.INSTANCE)) {
+			return getJsonFromVarchar(object, actualType);
+		}
+		return throwConstraintViolationException(actualType, this);
+	}
+
+	@Override
+	public boolean isCoercibleTo(
+			@SuppressWarnings("rawtypes") PDataType targetType) {
+		return equalsAny(targetType, this, PVarchar.INSTANCE);
+
+	}
+
+	@Override
+	public boolean isCoercibleTo(
+			@SuppressWarnings("rawtypes") PDataType targetType, Object value) {
+		return isCoercibleTo(targetType);
+	}
+
+	@Override
+	public boolean isSizeCompatible(ImmutableBytesWritable ptr, Object value,
+			@SuppressWarnings("rawtypes") PDataType srcType, Integer maxLength,
+			Integer scale, Integer desiredMaxLength, Integer desiredScale) {
+		if (ptr.getLength() != 0 && maxLength != null
+				&& desiredMaxLength != null) {
+			return maxLength <= desiredMaxLength;
+		}
+		return true;
+	}
+
+	@Override
+	public boolean isFixedWidth() {
+		return false;
+	}
+
+	@Override
+	public int estimateByteSize(Object o) {
+		PhoenixJson phoenixJson = (PhoenixJson) o;
+		return phoenixJson.toString().length();
```

Comment:
Add a estimateByteSize() on PhoenixJson as you don't want to have to string-ify the JSON just to get an estimate of the size.